dali/core/mm/composite_resource_test.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <type_traits>
+#include "dali/core/mm/composite_resource.h"
+#include "dali/core/mm/async_pool.h"
+#include "dali/core/mm/malloc_resource.h"
+
+namespace dali {
+namespace mm {
+
+static_assert(std::is_convertible<
+    decltype(make_composite_resource(
+        std::make_shared<async_pool_resource<memory_kind::device>>(), 42, "the answer"))*,
+    async_memory_resource<memory_kind::device>*>::value,
+    "A composite resource made from an async resource should have an async interface");
+
+static_assert(std::is_convertible<
+    decltype(make_shared_composite_resource(
+        std::make_shared<malloc_memory_resource>(), "the answer", "is", 42).get()),
+    memory_resource<memory_kind::host>*>::value,
+    "A composite resource made from an sync resource should have a memory_resource interface");
+
+static_assert(!std::is_convertible<
+    decltype(make_composite_resource(
+        std::make_shared<malloc_memory_resource>(), 42, "the answer"))*,
+    async_memory_resource<memory_kind::host>*>::value,
+    "A composite resource made from an sync resource should not have an async interface");
+
+namespace test {
+
+namespace {
+
+struct IsAlive {
+  IsAlive() {
+    alive = true;
+  }
+  ~IsAlive() {
+    alive = false;
+  }
+  bool alive;
+};
+
+
+template <memory_kind kind>
+struct DummyResource : public async_memory_resource<kind> {
```

Comment:
Would it make any sense to test `memory_resource` as well?

Reply:
It already implements `memory_resource` and the implementation is common to both interfaces (see `CompositeResourceBase`), so it's effectively tested.